### PR TITLE
SoftAssert#fail swallows actual root cause

### LINKED ISSUE #1778: SoftAssert#fail swallows actual root cause
### TestNG Version
6.14.3

### Expected behavior
SoftAssert#fail(final String message, final Throwable realCause) prints realCause trace

### Actual behavior
SoftAssert#fail(final String message, final Throwable realCause) doesn't reveal realCause

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ x ] IntelliJ
- [ ] NetBeans

### Test case sample

```java
import org.testng.Assert;
import org.testng.annotations.Test;
import org.testng.asserts.SoftAssert;

import java.io.PrintWriter;
import java.io.StringWriter;

public class AssertFailTest {

    private static final String HIDDEN_MSG = "hidden msg";
    private Exception hiddenExc = new Exception(HIDDEN_MSG);

    private void assertThrowableContainsText(Throwable hard, String text) {
        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        hard.printStackTrace(pw);
        String stackTrace = sw.toString();
        Assert.assertTrue(stackTrace.contains(text));
    }

    @Test
    public void testRootCauseInHardFail() {
        try {
            Assert.fail("default assert", hiddenExc);
        } catch (AssertionError hard) {
            assertThrowableContainsText(hard, HIDDEN_MSG);
        }
    }

    @Test
    public void testRootCauseInSoftFail() {
        try {
            SoftAssert s = new SoftAssert();
            s.fail("soft assert with missing root cause", hiddenExc);
            s.assertAll();
        } catch (AssertionError soft) {
            assertThrowableContainsText(soft, HIDDEN_MSG);
        }
    }
}
```

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #1778

Fixes #1778  .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

**Review by juherr — APPROVED** [review]

File: src/main/java/org/testng/asserts/SoftAssert.java [hunk 1]
```diff
@@ -38,7 +39,12 @@ public void assertAll() {
           sb.append(",");
         }
         sb.append("\n\t");
-        sb.append(ae.getKey().getMessage());
+        AssertionError error = ae.getKey();
+        sb.append(error.getMessage());
```
Comment (dr29bart) [inline src/main/java/org/testng/asserts/SoftAssert.java]: Value of ae map entry is never used.  Does it make sense to simplify?

File: src/main/java/org/testng/asserts/SoftAssert.java [hunk 2]
```diff
@@ -38,7 +39,12 @@ public void assertAll() {
...
+        Throwable cause = error.getCause();
+        if (cause != null) {
+          sb.append(" ").append(cause.getMessage());
```
[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/main/java/org/testng/asserts/SoftAssert.java]: @dr29bart - I didn't quite get that. What exactly do you mean by that? Are you talking about recursively invoking `error.getCause()` until we see a `null` root cause ?

File: src/main/java/org/testng/asserts/SoftAssert.java (hunk 1 above)
Comment (krmahadevan) [reply src/main/java/org/testng/asserts/SoftAssert.java]: Fixed.

**Review by dr29bart — APPROVED** [review]

File: src/main/java/org/testng/asserts/SoftAssert.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dr29bart) [reply src/main/java/org/testng/asserts/SoftAssert.java]: exactly

**Comment by krmahadevan** [comment]
ping @cbeust - Can you please help merge if you are fine with the changes

**Pull request merged**